### DISCUSSION TITLE
Equilibrium calculations help

### DISCUSSION THREAD

**Discussion by cummingsethan** [pycalphad/pycalphad#521] [Q&A]
Hi Folks, im working on using pycalphad for my masters and plotting charges/site fractions for a perovskite LMO and running into some errors  with equilibrium and could use some help.  I wanted to share what I've made and also see if y'all had any ideas for improvements:

conditions setting compared to thermocalc (cant set both v.X(LA) and v.X(MN), only 1 of them. would be nice to set both compositions)

at certain temps there are errors (see images), they all seem to be around -3e5 MU, but I can't find any rhyme or reason to why its happening in some cases but not all. Setting v.X('LA) = 0.2  instead of v.X('MN') also introduces significant gaps in the data plots.

I'm having issues with trying to use v.MU('O2') in equilibrium (list index is out of range)

If anyone has any suggestions for things to implement or try, that would be greatly appreciated.
![image](https://github.com/pycalphad/pycalphad/assets/160069519/fb1ab374-dc52-45f9-9c72-972250774f3a)
![image](https://github.com/pycalphad/pycalphad/assets/160069519/4ee9f0e1-1d90-4171-a516-2617b41b78e9)
![image](https://github.com/pycalphad/pycalphad/assets/160069519/2b62cb3c-387c-4316-ae44-cf2fcf5e5959)
![image](https://github.com/pycalphad/pycalphad/assets/160069519/9bff9f04-0b08-4a81-b432-e5efdee2fbe1)
![image](https://github.com/pycalphad/pycalphad/assets/160069519/db0b5f6f-ff47-48d2-8221-72b38f1ac563)

here is my set up process:

dbf=Database('lmo.tdb')
comps = ['LA', 'MN', 'O', 'VA']

desired_phase = 'PERV'
mod = Model(dbf, comps, desired_phase)
mod1= {desired_phase: mod}

conds = {v.P: 1e5, v.N: 1, v.X('MN'):0.2 , v.MU('O'):(-5e5, 1e5, 1e4), v.T: 1000}

eq= equilibrium(dbf,comps, desired_phase, conds, output = 'MU', model=mod1)

    chem_pot = masked_eq_res.MU.broadcast_like(masked_eq_res.Y).squeeze()
    y_data = masked_eq_res.Y.squeeze()
 

    # If chem_pot has an extra dimension compared to y adjust last one
    if chem_pot.ndim > y_data.ndim:
        chem_pot1 = chem_pot[..., 2]  # Adjust this number for [len(comps) - 2]
       
    data = np.column_stack((chem_pot1, y_data))
    
    data = data[:, ~np.isnan(data[0])]
    x = data[:, 0]  # First column as x values
    y = data[:, 1:]  # Remaining columns as y values
    ax.plot(x, y, label=label_name)

**Comment by bocklund** [comment]
Thanks for the question and the debugging you've done up to this point. Two questions:

1. Which version of pycalphad are you using? From a Python interpreter: `import pycalphad; print(pycalphad.__version__)`
2. Which conditions are you wanting to use that you're using in Thermo-Calc? You can only have c+2 conditions so for your ternary La-Mn-O system, N, P, T, X(MN), and M(O) fully specifiy the conditions

**Comment by cummingsethan** [comment]
0.10.4.post1

further debugging: x(mn): 0.1999 eliminates the issues. I'm not sure if it's a .tdb issue or a calculation issue for that front. I'm relatively newer to python as well so not sure if this is just a common issue for some calculations.

My goal is trying to also plot acr(O2). If I can get mu(O2),  I could convert it, but I am stuck on debugging the mu(O2) list indexing

IndexError                                Traceback (most recent call last)
Cell In[6], line 23
comps = ['LA', 'MN', 'O2', 'VA']

     19 mod1= {desired_phase: mod}
     21 conds = {v.P: 101325, v.N:1, v.X('MN'):.19999 , v.MU('O2'):(-5e5, 1e5, 0.7e4), v.T: 1000}
---> 23 eq= equilibrium(dbf,comps, desired_phase, conds, output = 'MU', model=mod)
  
     28 for sf_idx, sf_sym in enumerate(mod.site_fractions):

File ~\anaconda3\Lib\site-packages\pycalphad\core\equilibrium.py:281, in equilibrium(dbf, comps, phases, conditions, output, model, verbose, broadcast, calc_opts, to_xarray, scheduler, parameters, solver, callables, phase_records, **kwargs)
    279 coord_dict['component'] = pure_elements
    280 properties = starting_point(conds, state_variables, phase_records, grid)
--> 281 properties = _solve_eq_at_conditions(properties, phase_records, grid,
    282                                      list(str_conds.keys()), state_variables,
    283                                      verbose, solver=solver)
    285 # Compute equilibrium values of any additional user-specified properties
    286 # We already computed these properties so don't recompute them
    287 output = sorted(set(output) - {'GM', 'MU'})

File pycalphad/core/eqsolver.pyx:225, in pycalphad.core.eqsolver._solve_eq_at_conditions()

File pycalphad/core/eqsolver.pyx:103, in pycalphad.core.eqsolver.add_nearly_stable()

IndexError: list index out of range

Thank you!
